### DISCUSSION TITLE
Proposal for the new schedule and topics of part 3

### DISCUSSION THREAD

**Discussion by pl-marasco** [pangeo-data/pangeo-openeo-BiDS-2023#11] [Ideas]
DISCLAIMER: This is just a proposal and it's my understanding of the outcome of the last meeting. Everybody feels free to open a polite and open-minded discussion about all the aspects of the scheduling.

**Part-3: Unlocking the Power of Space Data with Pangeo & OpenEO**

- 14:00 Introduction to the afternoon session (... minutes)

*Data discoverability and searchability
- An overview of STAC (... minutes) 
- Overview of different sources and platforms

Different data exploitation approaches
- Hands-on "The Easy-peasy life with OpenEO" (... minutes)
    - Snow coverage example
- How to exploit data on Pangeo
    - Snow coverage pure xarray version 

Data and pre-processing general knowledge
- Cloud-friendly access to archival data with Kerchunk (... minutes)
- Introduction to chunking (... minutes)

Parallelization of the process 
- Scaling with Dask (... minutes)
- Scaling with OpenEO (how it works underneath) (... minutes)

How to go beyond 
- Understanding how to create a custom algorithm:  UDF (OpenEO) and ufunc (Xarray) (... minutes) 
- Common workflow that combines the best of the two “worlds” (... minutes)
- 17:45 Wrap-up and feedback survey (15 minutes)

@clausmichele Unfortunately I don't have all the GitHub names of the OpenEO group, could you please point this discussion as well to your community? 

@tinaok @acocac @annefou

**Comment by tinaok** [comment]
Thank you @pl-marasco, this looks good for me,
@clausmichele can you please confirm this is ok with you so that we can start to create the table of contents, place holders

**Reply by bschumac** [reply]
Part-3: Unlocking the Power of Space Data with Pangeo & OpenEO

- 14:00 Introduction to the afternoon session (5 minutes) -> suggestion @aljacob / @annefou 

*Data discoverability and searchability

- An overview of STAC and different sources and platforms (openeo.cloud, STAC browser, STAC Index ...) (15 minutes) -> @m-mohr 

14:30 Data and pre-processing general knowledge

- Introduction to chunking examples with netcdf, zarr and Kerchunk (30 minutes) -> @tinaok 
- Parallelization with Dask (30 minutes) -> @tinaok 

16:00 Different data exploitation approaches

- How to exploit data on OpenEO (30 minutes) -> @clausmichele 
        Snow coverage example
- How to exploit data on Pangeo (30 minutes) -> @pl-marasco 
        Snow coverage pure xarray version

17:00 How to go beyond

- Understanding how to create a custom algorithm: UDF (OpenEO) and ufunc (Xarray) (20 minutes)  -> @jdries
- Scaling with OpenEO (how it works underneath) (30 minutes) -> @jdries 
    
17:45 Wrap-up and feedback survey (15 minutes) @bschumac /  @annefou

**Reply by tinaok** [reply]
We have prepared https://github.com/pangeo-data/pangeo-openeo-BiDS-2023/blob/main/tutorial/part3/chunking_introduction.ipynb
Which will talk about chunking_zarr_kerchunk.  
And I propose that as the material for
Data and pre-processing general knowledge

We have also prepared  https://github.com/pangeo-data/pangeo-openeo-BiDS-2023/blob/main/tutorial/part3/scaling_dask.ipynb (see updated version in merge request)

That is for 
Parallelization of the process
Scaling with Dask

**Reply by bschumac** [reply]
Global feedback question? -> discuss with beginning questionnaire

**Comment by acocac** [comment]
@pl-marasco the proposed outline for the part 3 looks good to me too. 

@tinaok and I have opened PR #14 that restructures the Jupyter book according to Part 1 and 2 from the README, and part 3 proposed in this discussion. We've opened an accompanying issue #15 to validate which sections require to be completed.

**Comment by clausmichele** [comment]
@tinaok I can confirm for my snow workflow part, but I need a feedback from VITO for the parts where thay are also involved.
@soxofaan and @Pratichhya could you ask Jeroen (invited him to this repo) about these parts: **Scaling with OpenEO (how it works underneath) (... minutes)** and **Understanding how to create a custom algorithm: UDF (OpenEO) and ufunc (Xarray) (... minutes)** ?

**Reply by tinaok** [reply]
@clausmichele thank you.  Who will be contributing to 'Data discoverability and searchability' part from openeo side? would you mind tag the person?

**Reply by clausmichele** [reply]
Tomorrow morning at 10 we have an internal meeting and define all the responsible persons for each part.

**Reply by Pratichhya** [reply]
@tinaok, could you please add https://github.com/jdries too, to this repository?

**Reply by tinaok** [reply]
Invitations to @jdries @bschumac were sent before, it is 'awaiting response'

**Reply by soxofaan** [reply]
> @soxofaan .. could you ask Jeroen ...

Sorry,  the notifications for this repo apparently end up in another mailbox I'm used to, so I missed this. I assume Jeroen is already in the loop by now?

**Comment by jdries** [comment]
@clausmichele  
Scaling with openEO is actually something I'll have to work out still. I'm guessing it's simply going to be slide based, maybe with some concrete cases like continental mapping.
For UDF's: we now also put them in part 2 I believe, but we could show another case, or dive into it a bit deeper.

**Comment by clausmichele** [comment]
@tinaok @acocac @pl-marasco the EODC Team will take care about this part **An overview of STAC (... minutes)**, the length will be around 15/20 minutes. The content (to be finalized, it will be posted here) will be more or less the following:

-	What kind of problem is STAC trying to solve, main concepts
-	What are the main components (Catalog, Collection, Item, Asset)
-	With a public Catalog, show a sample of Collection, Item, Asset as JSON (Firefox renders it nicely)
-	Show the STAC Browser referring to the same Collection/Items/Assets we showed before
-	Mention that we support it in openEO
-	Mention the extensions (raster, etc), useful for interoperability with third party libraries.
-	Open Source, open to new extensions.

**Reply by jdries** [reply]
Key takeaways for this IMO:
- not any 'valid' STAC is also a 'usable' STAC, the openEO perspective is the topic of this PR: https://github.com/openEOPlatform/documentation/pull/55/files
- link to FAIR principles, to understand the motivation for creating quality metadata

And of course it's nice to show how there's this nice synergy between tools in the STAC ecosystem, like openEO batch job output being visualized in the STAC browser:
![image](https://github.com/pangeo-data/pangeo-openeo-BiDS-2023/assets/5937096/9933510d-dc68-4c26-9a19-cf29f6ec607d)

**Reply by m-mohr** [reply]
Link to the STAC slides: https://docs.google.com/presentation/d/1S5QeRNGwfRSLGrw0w0wIR90nIYxhvimoIO97vqMZs4M/edit?usp=sharing

Open for feedback.

**Comment by jdries** [comment]
I'm still shuffling around my content a bit, trying to find a good flow:

In part 2, I have 30minutes for 'advanced workflows', that I can probably fill with this:

- complex workflows with predefined functions (worldwater, rank composites)
- FAIR metadata generation
- Short note on costs: https://docs.openeo.cloud/federation/accounting.html#platform-credit-rates

In part 3, 
Advanced workflows: show a UDF example

In the 'under the hood/scaling' part, I can then discuss:
- continental mapping
- an actual 'under the hood' look of a backend

I also have some 'orphan' topics that could be used to fill up the agenda if nobody else is mentioning them:
- dataset sampling (like for ML training data extraction) -> there's actually a nice link to make with local processing/Pangeo here
- zonal statistics (at large scale)

**Reply by tinaok** [reply]
thanks @jdries if the notebook that you might show in udf example or orphan part exist somewhere I can have a look and see if I can link that with chunking/dask/apply_u_func example in part3?

**Reply by jdries** [reply]
There's already one case in the notebook of that part:
https://github.com/pangeo-data/pangeo-openeo-BiDS-2023/blob/main/tutorial/part3/advanced_workflow.ipynb
which is about parcel delineation.

We however also have this nice example of peak/valley detection that uses apply_ufunc already, and has notebooks in openEO and pure XArray:
https://github.com/Open-EO/FuseTS/blob/main/src/fusets/peakvalley.py#L38
https://github.com/Open-EO/FuseTS/blob/main/notebooks/AI4FOOD_PeakValley_Detection_NDVI_S2/peakvalley_detection.ipynb

If you think these are useful, I can try running them tomorrow to double check if they work as expected.

**Reply by tinaok** [reply]
Thanks, @jdries, for the excellent notebooks! I was wondering how we can connect these notebooks with the snow coverage example from your team ; 
Hands-on "The Easy-peasy life with OpenEO" (... minutes)
Snow coverage example 
Any idia?

**Reply by jdries** [reply]
Hi Tina,
I had a look, but don't really have a quick and easy solution. We can of course run peak valley detection on the snow timeseries, but not sure if that is something meaningfull to do.
The other example with parcel delineation is even a bit more focused on agriculture. Of course you can imagine running a delineation algorithm over a snow area as well, but the model we trained is mostly tested on agriculture.

So not sure what to do, but perhaps it's also not bad that we show examples covering multiple thematic areas?

**Reply by acocac** [reply]
@jdries 

fyi, I've tested the notebooks in Pangeo-EOSC using my OpenEO credentials. Please find my findings below.

- Parcel delineation: it works. I understand the notebook is mostly inference, however it'd be great to let users know how to generate the interim plots i.e. raw/post-processed predictions.
  - missing file, `resources/training_parcels.geojson` to run the last cell. 
- Peak Valley: also a great notebook! I don't have errors submitting jobs, but some issues exists in the plotting steps.
   - requires to install `vam`, `vam.whittaker` and `jupyter_bokeh` packages. All are pip installable. Note, after restarting the kernel I'm having some issues when importing `vam.whittaker`: 
       - `ImportError: cannot import name 'ws2d' from 'vam.whittaker' (unknown location)`
   - add a cell to use a relative import to access functions within `src/fusets`
   - error in the last cell when generating the interactive plotting: Failed to load model class `BokehModel`

I hope above assessment is useful.

**Reply by jdries** [reply]
That's very useful indeed! Seems easy enough to fix. 
(PS I'm mostly working in the evenings this week, and probably on thursday.)

**Reply by jdries** [reply]
@acocac for peak-valley, I think it's mostly a matter of doing `pip install fusets`. I can try to add it to the environment, or we can simply leave it out. Some of these libraries like whittaker are not so good at managing backwards compatibility and thus cause trouble downstream.

**Reply by acocac** [reply]
@jdries - I was importing `fusets` locally from the repository. I'll try it by installing fusets so it could handle better the dependencies.

**Reply by Pratichhya** [reply]
@acocac I tried to replicate your issue. Though I didn't get `ImportError: cannot import name 'ws2d' from 'vam.whittaker' (unknown location)` , I noticed that when we separately install vam.whittaker, there is a version collision causing several errors. If it helps, replicating the environment shared in this https://github.com/Open-EO/FuseTS/blob/main/workshops/BIDS/workshop/infrastructure/Dockerfile maybe you can try.

Also, @jdries, I think https://github.com/Open-EO/FuseTS/blob/main/notebooks/Local/FuseTS%20-%20Peak%20Valley%20Detection.ipynb is the latest notebook inside the local directory that also has initial steps on installing jupyter_bokeh and jupyter-widget. @acocac could you please try with this, if this resolves your plot issue?

I managed to get the plot, but its dynamic nature is throwing some errors when clicked, I will make a push once solved.

**Reply by Pratichhya** [reply]
Hi @acocac,

Could you please check if https://github.com/Open-EO/FuseTS/blob/main/notebooks/Local/FuseTS%20-%20Peak%20Valley%20Detection_holoviews.ipynb plots works for you?
There seems to be a bug in holoview. Though I can create it in my local environment, it requires some installation when tested in a new environment.

 I would also recommend not to separately install vam.whittaker, there could be wheel version conflict. Therefore use https://github.com/Open-EO/FuseTS/blob/peak_valley_notebook/environment_whittaker.yml

**Reply by acocac** [reply]
@Pratichhya I'm on holidays today. I'd suggest that @tinaok @pl-marasco @annefou help to test it today if urgent. I'll try to have a look at the issue this weekend.

**Reply by Pratichhya** [reply]
Sure sure it's ok @acocac. Have a nice holiday 😄

**Reply by pl-marasco** [reply]
@Pratichhya first I will try on my local machine and then test on PangeoEOSC deployment. I'll get back to you soon

**Reply by pl-marasco** [reply]
@Pratichhya despite all my tests I'm facing the same issue as @acocac 

In all the tests I've conducted the issue is always linked to the vam-whittaker. I've been testing your yml file for the conda but it doesn't work. 
The only way to get a usable env is through docker 
Forcing installation without dependencies didn't work as well

**Comment by tinaok** [comment]
Does anyone knows where can I find the link to snow coverage example? @pl-marasco @acocac ?
I updated chunking and dask notebook, but the datas I can update that to example which we will use before the chunking-dask (i.e. snow) or after (related to UDF / apply u_func example)

**Reply by clausmichele** [reply]
@tinaok the example using openEO + STAC + xarray + dask is this one: https://github.com/EO-College/cubes-and-clouds/blob/main/lectures/3.1_data_processing/exercises/_alternatives/31_data_processing_stac.ipynb
There's not yet a version without openEO. I know that @pl-marasco said he was trying to convert it but didn't hear back from him recently.

**Reply by pl-marasco** [reply]
@clausmichele sorry to disappear from the radar for a little while but, afert quite some tests, I had I had to abbandon the idea of using the CDSE as data source. 
The coonverted notebook relay on the Element84/AWS source and is almost ready (hope to release it on Monday).
The only adaptation that I had too do is the time window that now is focus one year later then the original.

**Comment by jdries** [comment]
Note that I already made a commit to actually only import the whittaker dependency when it's actually used. Problem is, I'm not sure how to do a bugfix release right now.
So checking out from source would be one way to at least fix it, the other is to simply take the small piece of code out of FuseTS, and use it without all the other dependencies.

________________________________

**Reply by jdries** [reply]
FYI, I went for option 2 and copied the relevant code which does not have dependency issues. This also will make it a lot easier for users to play with themselves during the workshop.
